File: go/cmd/cadreen/commands/healing.go
package commands

import (
	"encoding/json"
	"fmt"

	"github.com/timothy-billingrails/cadreen-sdks/go/cmd/cadreen/output"

	cadreen "github.com/timothy-billingrails/cadreen-sdks/go/cadreen"

	"github.com/spf13/cobra"
)

var healingCmd = &cobra.Command{
	Use:   "healing",
	Short: "Self-healing stats and diagnostics",
	Long:  `View healing statistics, precedents, and diagnose failures.`,
}

var healingStatsCmd = &cobra.Command{
	Use:   "stats",
	Short: "Healing statistics",
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/healing/stats", nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var stats cadreen.HealingStatsResponse
		if err := json.NewDecoder(resp.Body).Decode(&stats); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(stats, format)
		default:
			printHealingStats(stats)
		}

		return nil
	},
}

var healingPrecedentsCmd = &cobra.Command{
	Use:   "precedents",
	Short: "List healing precedents",
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/healing/precedents", nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var result struct {
			Precedents []cadreen.HealingPrecedent `json:"precedents"`
			Count      int                        `json:"count"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(result, format)
		case output.FormatVerbose:
			printPrecedentsVerbose(result.Precedents)
		default:
			printPrecedentsText(result.Precedents)
		}

		return nil
	},
}

var healingDiagnoseCmd = &cobra.Command{
	Use:   "diagnose [error]",
	Short: "Diagnose a failure",
	Long: `Diagnose why something failed and get recovery suggestions.

Examples:
  cadreen healing diagnose "connection timeout to stripe API"
  cadreen healing diagnose "403 forbidden" --tool stripe_refund`,
	Args: cobra.MinimumNArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		toolName, _ := cmd.Flags().GetString("tool")

		body := map[string]string{"error_message": args[0]}
		if toolName != "" {
			body["tool_name"] = toolName
		}

		resp, err := rawDo("POST", "/api/v1/cadreen/healing/diagnose", body)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var diag cadreen.HealingDiagnosis
		if err := json.NewDecoder(resp.Body).Decode(&diag); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(diag, format)
		default:
			printDiagnosis(diag)
		}

		return nil
	},
}

func init() {
	healingDiagnoseCmd.Flags().String("tool", "", "tool name that failed")

	healingCmd.AddCommand(healingStatsCmd)
	healingCmd.AddCommand(healingPrecedentsCmd)
	healingCmd.AddCommand(healingDiagnoseCmd)
	rootCmd.AddCommand(healingCmd)
}

func printHealingStats(stats cadreen.HealingStatsResponse) {
	fmt.Println("Healing Statistics")
	fmt.Println()
	if stats.TotalPrecedents > 0 {
		fmt.Printf("  Total precedents: %d\n", stats.TotalPrecedents)
	}
	if stats.SuccessfulRecoveries > 0 {
		fmt.Printf("  Successful: %d\n", stats.SuccessfulRecoveries)
	}
	if stats.FailedRecoveries > 0 {
		fmt.Printf("  Failed: %d\n", stats.FailedRecoveries)
	}
	if stats.SuccessRate > 0 {
		fmt.Printf("  Success rate: %.0f%%\n", stats.SuccessRate*100)
	}
}

func printPrecedentsText(precedents []cadreen.HealingPrecedent) {
	if len(precedents) == 0 {
		fmt.Println("No healing precedents.")
		return
	}

	fmt.Printf("Healing precedents (%d):\n\n", len(precedents))
	for _, p := range precedents {
		status := "✗"
		if p.Success {
			status = "✓"
		}
		fmt.Printf("  %s %s\n", status, p.ErrorType)
		if p.ToolName != "" {
			fmt.Printf("    Tool: %s\n", p.ToolName)
		}
		if p.RecoveryStrategy != "" {
			fmt.Printf("    Strategy: %s\n", p.RecoveryStrategy)
		}
		fmt.Println()
	}
}

func printPrecedentsVerbose(precedents []cadreen.HealingPrecedent) {
	for _, p := range precedents {
		fmt.Printf("ID:        %s\n", p.ID)
		fmt.Printf("Error:     %s\n", p.ErrorType)
		fmt.Printf("Success:   %v\n", p.Success)
		fmt.Printf("Attempts:  %d\n", p.Attempts)
		if p.ToolName != "" {
			fmt.Printf("Tool:      %s\n", p.ToolName)
		}
		if p.RecoveryStrategy != "" {
			fmt.Printf("Strategy:  %s\n", p.RecoveryStrategy)
		}
		if p.WhatWorked != "" {
			fmt.Printf("Worked:    %s\n", p.WhatWorked)
		}
		fmt.Println()
	}
}

func printDiagnosis(diag cadreen.HealingDiagnosis) {
	fmt.Println("Diagnosis")
	fmt.Println()
	if diag.ErrorCategory != "" {
		fmt.Printf("  Category:    %s\n", diag.ErrorCategory)
	}
	if diag.RootCause != "" {
		fmt.Printf("  Root cause:  %s\n", diag.RootCause)
	}
	if diag.SemanticReason != "" {
		fmt.Printf("  Reason:      %s\n", diag.SemanticReason)
	}
	if diag.CanRetry {
		fmt.Println("  Can retry:   yes")
	}
	if diag.NeedsHuman {
		fmt.Println("  Needs human: yes")
	}
	if diag.Confidence > 0 {
		fmt.Printf("  Confidence:  %.0f%%\n", diag.Confidence*100)
	}
}

File: python/cadreen/resources/healing.py
from __future__ import annotations

from typing import Any

from ..client import HttpClient
from ..types import (
    HealingStatsResponse,
    ListHealingPrecedentsResponse,
    HealingDiagnosis,
    HealingPrecedent,
    StrategyCount,
    ToolHealingStats,
    TimeRange,
    Pagination,
)


class HealingResource:
    def __init__(self, client: HttpClient) -> None:
        self._client = client

    async def stats(self) -> HealingStatsResponse:
        raw = await self._client.get("/api/v1/cadreen/healing/stats")
        strategies = None
        if raw.get("common_strategies"):
            strategies = [StrategyCount(strategy=s["strategy"], count=s["count"]) for s in raw["common_strategies"]]
        top_tools = None
        if raw.get("top_tools"):
            top_tools = [
                ToolHealingStats(
                    tool_name=t["tool_name"],
                    total=t["total"],
                    successful=t["successful"],
                    failed=t["failed"],
                    success_rate=t["success_rate"],
                    top_strategy=t.get("top_strategy"),
                )
                for t in raw["top_tools"]
            ]
        time_range = None
        if raw.get("time_range"):
            tr = raw["time_range"]
            time_range = TimeRange(first_precedent=tr.get("first_precedent"), last_precedent=tr.get("last_precedent"))
        return HealingStatsResponse(
            total_precedents=raw.get("total_precedents"),
            successful_recoveries=raw.get("successful_recoveries"),
            failed_recoveries=raw.get("failed_recoveries"),
            success_rate=raw.get("success_rate"),
            avg_duration_ms=raw.get("avg_duration_ms"),
            common_strategies=strategies,
            top_tools=top_tools,
            by_category=raw.get("by_category"),
            time_range=time_range,
        )

    async def precedents(self) -> ListHealingPrecedentsResponse:
        raw = await self._client.get("/api/v1/cadreen/healing/precedents")
        precedents = [
            HealingPrecedent(
                id=p["id"],
                error_type=p["error_type"],
                success=p["success"],
                attempts=p.get("attempts", 0),
                confidence=p.get("confidence", 0.0),
                tool_name=p.get("tool_name"),
                error_category=p.get("error_category"),
                semantic_reason=p.get("semantic_reason"),
                root_cause=p.get("root_cause"),
                recovery_strategy=p.get("recovery_strategy"),
                what_worked=p.get("what_worked"),
                what_failed=p.get("what_failed"),
                duration_ms=p.get("duration_ms"),
                created_at=p.get("created_at"),
                domain=p.get("domain"),
                tags=p.get("tags"),
            )
            for p in raw.get("precedents", [])
        ]
        pagination = None
        if raw.get("pagination"):
            p = raw["pagination"]
            pagination = Pagination(limit=p["limit"], offset=p["offset"], has_more=p["has_more"])
        return ListHealingPrecedentsResponse(
            precedents=precedents,
            count=raw.get("count", 0),
            pagination=pagination,
        )

    async def diagnose(
        self,
        error_message: str,
        *,
        tool_name: str | None = None,
        trace_id: str | None = None,
    ) -> HealingDiagnosis:
        body: dict[str, Any] = {"error_message": error_message}
        if tool_name is not None:
            body["tool_name"] = tool_name
        if trace_id is not None:
            body["trace_id"] = trace_id
        raw = await self._client.post("/api/v1/cadreen/healing/diagnose", body)
        return HealingDiagnosis(
            error_category=raw.get("error_category"),
            semantic_reason=raw.get("semantic_reason"),
            root_cause=raw.get("root_cause"),
            can_retry=raw.get("can_retry"),
            needs_sub_execution=raw.get("needs_sub_execution"),
            needs_human=raw.get("needs_human"),
            should_skip=raw.get("should_skip"),
            needs_re_decide=raw.get("needs_re_decide"),
            needs_try_alternative=raw.get("needs_try_alternative"),
            retry_delay_ms=raw.get("retry_delay_ms"),
            confidence=raw.get("confidence"),
        )
